### ISSUE TITLE
Ability to use context in page margin attribute

### ISSUE THREAD

**Issue by ramkumarkb** [typst/typst#5182]
### Description

I am logging this issue here as a follow up from the discussion in the Forum, and answered by @laurmaedje - 

https://forum.typst.app/t/how-to-get-differently-sized-header-or-footer-depending-on-page-number/322 

Today we cannot use context in the margin attribute of the page : 
```
#set page(
  columns: 1,
  paper: "a4",
  flipped: false,
  // introspection APIs are not allowed in margin 
  margin: context { 
    (x: 16mm, y: 27.5mm),
    if counter(page).get() == counter(page).final() {
       (x: 26mm, y: 37.5mm)
    }
  // introspection APIs allowed for header & footer
  header: context { ... }
  footer: context { ... }
  ```

### Use Case

In the Forum, it is explained that changing the side margins (x) is difficult. However the more common usecase is to change the top / bottom margin (y). 

Typically, one needs to change these margins for the cover or the last page, especially the top / bottom.

**Comment by th1j5** [comment]
In addition, one cannot calculate the margins based on the page dimensions now.
When using some sort of formula to calculate the margins based on the page dimensions, using a relative length is not enough.
See for example https://styleguide.ugent.be/basic-principles/grid-and-layout.html.

**Comment by tyler-wright** [comment]
I found this page while trying to recreate my law firm letterhead in Typst. Unfortunately, because of this limitation, it is not possible to recreate it and have it function the way that you would expect it to (for instance, the way it functions in Microsoft Word).

Seems like a strange feature to be essential for adoption in the legal profession, but I promise you it is. People need to be able to recreate their letterheads perfectly. It's a little bit of a point of pride.

I'm gonna have a look into your contribution guidelines and maybe run a few AI queries over your codebase to unpack what's involved in putting together a PR. I would like to contribute to a project like this because it is very impressive.

**Comment by laurmaedje** [comment]
@tyler-wright Unfortunately, a comprehensive solution is complex. I'm optimistic that we'll resolve this in time, but it's not something that could easily be PRed.

Depending on your exact requirements, it may be possible to work around your limitation. Might be worth posting your exact requirements on https://forum.typst.app/.

**Comment by tyler-wright** [comment]
No worries @laurmaedje - I will post the requirements for you on the forum because I think it will be helpful when you do get around to it. I can demonstrate use cases for both the X and Y margin variations across different pages based on real world scenarios that you can add to the ones you have above (the style guide example is fairly good though).

I'm just exploring workarounds now so I'll probably get something in place so I can adopt the system as an early adopter. If I do I will post there along with a full description of my requirements.

Thanks so much for your reply.

**Comment by th1j5** [comment]
Apparently, I worked around it by `set page(margin: 0pt)` and using `blocks`/`place`/`grid`... See [ugent-cover-layout-page](https://codeberg.org/th1j5/typst-sos-ugent-style/src/commit/f9617db6a811fc824d85c0b6c82c87187851e575/src/styling/cover.typ#L214-L222) and the function it calls (`ugent-cover-layout`). For just a title page, this is acceptable, but it probably isn't for 'normal' margins.

**Comment by tyler-wright** [comment]
No, you definitely haven't recreated it. The key part of this feature request is that you want to be able to flow text irrespective of what the content is across the different pages while maintaining the different margins, if that makes sense.

So for instance you might have on the x-axis and the first page 5cm margin on the left and then 5cm margin on the bottom creating a narrow column on the right hand side of the page. Then you can put your letter content in that column and then have it flow seamlessly onto the second page where it will then spread out to cover the entire page which might have normal margins.

The key difference with your solution is, I believe, that you can't put unlimited content into the first page. You at some point have to say, okay, this is where the content stops. It doesn't just keep flowing onto the other pages with different margin layouts. I'm about to try it now and if I have time I'll come back and mention it but I suspect this is the difference.

I did find a workaround for my case and it does work effectively but only on the x-axis margins nothing on the y-axis margins.

**Comment by Binarus** [comment]
Is there already a roadmap for that feature?

I'm also in the situation where I'd like to rebuild our company's letter structure in typst, which currently is not (easily) possible. I'm at 99%, but as usual, the last percent is the show stopper. I believe that our use case is not completely unimportant as a lot of people may be interested in it:

As everybody knows, a company that sends a letter (in electronic or printed form) is legally required to include a bunch of legal information. While law does not exactly dictate where to place that information, some common patters have evolved. The most prevalent method is to put the legal information (or a part of it) into the footer on every page. That makes great sense for all pages except for the first page.

On the first page, space is usually very limited, because the first page contains the address fields for the sender and the recipient, logo(s), the sender's contact information, maybe a special area denoting the subject, and so on.

At the same time, the very same content that takes up so much space on the first page already contains 90% of the legal information, and it would not make any sense to repeat these 90% in the footer (as on the other pages). This means that the footer (and thus, the bottom margin) can (and should!) be considerably smaller on the first page than on the other pages. This would give back some of the precious space on the first page.

As typst doesn't support dynamic margins at the moment, I'll try to work around the problem by setting the bottom margin to 0 and to simulate the margin by a floating box. Or something like that. Or perhaps footnotes come to rescue. I haven't looked into them yet, but they are usually at the end of a page, and the height of the area where they are placed must depend on their content. This should be worth some tests.

**Comment by laurmaedje** [comment]
@Binarus You could try something like this:
```typ
#set page(margin: (bottom: 10cm))
#place(bottom, float: true, block(height: -6cm))
#lorem(2000)
```
Setting the margin to your usual one, but fake-reducing it with a float on the first page.

(I don't think we currently test for this behavior, but it seems to work.)

**Comment by Binarus** [comment]
Thank you very much for that tip! I wouldn't have come to that idea, probably because I was searching for a generic mechanism that works not only on the first page.

In my case, your code does not work out of the box for the following reason: Although the footer on the first page is considerably smaller (in height) than on the other pages, it is not completely empty. With the code above, the footer's content on the first page is overlaid by the `#lorem` text when I fake-reduce the margin that way.

However, I believe that the approach you have shown will work. I intend to solve my problem by moving the footer's content on the first page down towards the page end by the same amount as the margin is fake-reduced. This should be possible because I already have code that changes the footer's contents depending on the page number.

I'll report whether this solves the problem.

**Comment by Binarus** [comment]
In the meantime, I had success.

As expected, I had to move the footer's content down (towards the page end) on the first page by the amount the margin got fake-reduced, but that was a no-brainer (a simple `#h()` before the actual footer content was sufficient).

But initially, I had a hard time trying to understand why your trick *somehow* worked, but the vertical gap between the `#lorem` text and the footer was unexpectedly big. To achieve the desired result, I had to set the floating block height to a negative value far beyond my expectation, which was a process of trial and error.

Fortunately, after a while, I remembered the `footer-descent` and `clearance` properties. From then on, I understood what is happening, and now I can easily change the look if necessary without fiddling with a caliper.

Thank you very much again for the great support! That's a very elegant solution to the problem. I hope that future versions won't break it (unless these versions provide official means to dynamically adjust margins depending on context or state).
